File: R/extract_spectra.R
## Extract spectra from json data downloaded from Biolegend's Spectra Analyzer
## See: https://www.biolegend.com/spectraanalyzer

wd = dirname(this.path::here())  # wd = '~/github/R/helperFlowCell'
library('rjson')
library('optparse')
library('logr')
import::from(file.path(wd, 'R', 'utils', 'file_io.R'),
    'join_many_csv', .character_only=TRUE)


# ----------------------------------------------------------------------
# Pre-script settings

# args
option_list = list(
    make_option(c("-i", "--input-dir"), default='ref/raw_spectra_json',
                metavar='ref/raw_spectra_json', type="character",
                help="specify the directory of json files"),

    make_option(c("-o", "--output-dir"), default="ref/raw_spectra",
                metavar="ref/raw_spectra", type="character",
                help="set the output directory for the data"),

    make_option(c("-t", "--troubleshooting"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="enable if troubleshooting to prevent overwriting your files")
)
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
troubleshooting = opt[['troubleshooting']]

# Start Log
start_time = Sys.time()
log <- log_open(paste0("extract_spectra-",
    strftime(start_time, format="%Y%m%d_%H%M%S"), '.log'))
log_print(paste('Script started at:', start_time))

# create output dir
if (!dir.exists(file.path(opt[['output-dir']]))) {
    dir.create(file.path(opt[['output-dir']]), recursive=TRUE)
}


# ----------------------------------------------------------------------
# Convert each file

# get files
filepaths <- list.files(file.path(wd, opt[['input-dir']]), full.names=TRUE)
log_print(paste('Number of files found:', length(filepaths)))

# convert
for (filepath in filepaths) {

    log_print( paste('Processing:', basename(filepath)) )
    filename_sans_ext <- tools::file_path_sans_ext(basename(filepath))
    
    # convert
    raw_data <- fromJSON(file=filepath)
    df <- as.data.frame(do.call(rbind, raw_data[['data']]))
    colnames(df) <- c('Wavelength', 'Intensity')

    # normalize
    max_intensity <- max(df[2])
    if (max_intensity > 100) {
        df[2] <- df[2] / max_intensity
    } else if (max_intensity > 1) {
        df[2] <- df[2] / 100
    }

    # save
    if (!troubleshooting) {
        write.table(
            df,
            file = file.path(opt[['output-dir']], paste0(filename_sans_ext, '.csv')),
            row.names = FALSE, sep=','
        )
    }
}


# ----------------------------------------------------------------------
# Combine output

log_print(paste('Combining...'))

filepaths <- list.files(file.path(wd, opt[['output-dir']]), full.names=TRUE)
combined <- join_many_csv(
    filepaths,
    index_cols='Wavelength', value_cols='Intensity',
    all.x=TRUE, all.y=TRUE
)

# remove half wavelengths
combined <- combined[(combined[['Wavelength']]%%1 < 0.01), ]

# save
if (!troubleshooting) {
    write.table(
        combined,
        file = file.path(dirname(opt[['output-dir']]), paste0('combined-spectra', '.csv')),
        row.names = FALSE, sep=',',
        na=""
    )
}

end_time = Sys.time()
log_print(paste('Script ended at:', Sys.time()))
log_print(paste("Script completed in:", difftime(end_time, start_time)))
log_close()

File: python/setup/extract_spectra.py
#!/usr/bin/env python3
"""Extract spectra from JSON data downloaded from Biolegend's Spectra Analyzer."""

import os
import sys
import json
import argparse
import logging
from datetime import datetime

import pandas as pd

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from tools.file_io import join_many_csv


def main():
    parser = argparse.ArgumentParser(description='Extract spectra from JSON')
    parser.add_argument('-i', '--input-dir', default='data/raw_spectra',
                        help='Input directory with JSON files')
    parser.add_argument('-o', '--output-dir', default='data/raw_spectra/output',
                        help='Output directory')
    parser.add_argument('-t', '--troubleshooting', action='store_true')

    args = parser.parse_args()

    # Setup
    wd = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

    # Logging
    start_time = datetime.now()
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')
    logging.info(f'Script started at: {start_time}')

    # Create output directory
    os.makedirs(os.path.join(wd, args.output_dir), exist_ok=True)

    # Get JSON files
    input_dir = os.path.join(wd, args.input_dir)
    filepaths = [os.path.join(input_dir, f) for f in os.listdir(input_dir)
                 if f.endswith('.json')]
    logging.info(f'Number of files found: {len(filepaths)}')

    # Convert each file
    for filepath in filepaths:
        logging.info(f'Processing: {os.path.basename(filepath)}')
        filename = os.path.splitext(os.path.basename(filepath))[0]

        with open(filepath) as f:
            raw_data = json.load(f)

        df = pd.DataFrame(raw_data['data'], columns=['Wavelength', 'Intensity'])

        # Normalize
        max_intensity = df['Intensity'].max()
        if max_intensity > 100:
            df['Intensity'] = df['Intensity'] / max_intensity
        elif max_intensity > 1:
            df['Intensity'] = df['Intensity'] / 100

        if not args.troubleshooting:
            df.to_csv(os.path.join(wd, args.output_dir, f'{filename}.csv'),
                      index=False)

    # Combine output
    logging.info('Combining...')
    output_files = [os.path.join(wd, args.output_dir, f)
                    for f in os.listdir(os.path.join(wd, args.output_dir))
                    if f.endswith('.csv')]

    if output_files:
        combined = join_many_csv(
            output_files, index_cols=['Wavelength'], value_cols=['Intensity'],
            all_x=True, all_y=True
        )

        # Remove half wavelengths
        combined = combined[combined['Wavelength'] % 1 < 0.01]

        if not args.troubleshooting:
            parent_dir = os.path.dirname(os.path.join(wd, args.output_dir))
            combined.to_csv(os.path.join(parent_dir, 'combined-spectra.csv'),
                            index=False)

    end_time = datetime.now()
    logging.info(f'Script ended at: {end_time}')
    logging.info(f'Script completed in: {end_time - start_time}')


if __name__ == '__main__':
    main()
